---
title: "Comparing community differences between a variable"
author: "Kim Vincent"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: word_document
---

```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Current Session Versions, include = TRUE}
sessionInfo()
```

```{r Install packages and load libraries, include = TRUE}
# options(repos = c(CRAN = "http://cran.rstudio.com"))
# 
# Install devtools
###################################################
if (!require(devtools)) {install.packages("devtools")}
library(devtools)

# Install packages
###################################################
if (!require(knitr)) {install.packages("knitr")}
if (!require(tidyverse)) {install.packages("tidyverse")}
devtools::install_github("leffj/mctoolsr")
devtools::install_github("kassambara/ggpubr")
if (!require(ggrepel)) {install.packages("ggrepel")} # makes labels repel data points.
if (!require(vegan)) {install.packages("vegan")} # uses PERMANOVA to test community differences.
if (!require(grid)) {install.packages("grid")}

# Load libraries
###################################################
library(knitr); packageVersion("knitr")
library(tidyverse); packageVersion("tidyverse") # for data manipulation and plotting
library(mctoolsr); packageVersion("mctoolsr")
library(beepr); packageVersion("beepr")
library(ggpubr); packageVersion("ggpubr")
library(ggrepel); packageVersion("ggrepel")
library(vegan); packageVersion("vegan") # for vegetation and community analysis
library(grid); packageVersion("grid")

```

Let's get organized! 
```{r, include = TRUE}
# set chapter
chapter <- "Chapter_2_DOM"
path <- "/Users/vivaeco/Documents/Boulder/Research/" # path to chapter

# set primer
primer <- "ITS" # 16S, ITS, ITS_16S, 18S
paste0("The primer we are examining is: ", primer)

# set variable of interest
var <- "DOMsource"
paste0("The variable we are comparing communities between is: ", var)
```

This chunk creates project names and folders based on the primers being examined. Even if I don't use the folders in this script, I like to keep this chunk in each script so my folders are set up uniformly.
```{r, include = TRUE}

# Defines the project automatically by the primer specified above. 
if (primer == "16S") {
project <- "16S"
} else if (primer == "18S") {
project <- "18S"
} else if (primer == "ITS_16S") {
project <- "NetworkCrossDomain"
} else if (primer == "ITS") {
project <- "ITS"
}

# set working directory
wd <- paste0(path, "/", chapter, "/", project)

# Create a filepath and Data folder within the working directory
data.fp <- file.path(paste0(wd, "/Data"))
if (!dir.exists(data.fp)) {
  dir.create(data.fp)
  }
list.files(data.fp) # List all files in folder to check path

# Create a filepath and Results folder within the working directory
results.fp <- file.path(paste0(wd, "/Results"))
if (!dir.exists(results.fp)) {
  dir.create(results.fp)
  }
list.files(results.fp) # List all files in folder to check path

# Create a filepath and Results folder within the working directory
commdiff.fp <- file.path(paste0(results.fp, "/CommDiff"))
if (!dir.exists(commdiff.fp)) {
  dir.create(commdiff.fp)
  }
list.files(commdiff.fp) # List all files in folder to check path

# Create a Figures folder in the Results folder within the working directory
figures.fp <- file.path(paste0(results.fp, "/Figures"))
if (!dir.exists(figures.fp)) {
  dir.create(figures.fp)
  }
list.files(figures.fp) # List all files in folder to check path
```

This chunk starts by inputting a rarefied mctoolsr object, defines the OTU table, and cleans up the files. 
```{r, include = TRUE}
# Load mctoolsr object
input_rare <- readRDS(paste0(data.fp, "/Objects/MctoolsrInputObj_Rare_", primer, ".rds")) 

# Define OTU table
OTU <- t(input_rare$data_loaded)

# Define mapping file/ metadata
metadata <- input_rare$map_loaded %>% 
  rownames_to_column(var = "SampleID")

# Pretty up the taxa table
input_rare$taxonomy_loaded <- input_rare$taxonomy_loaded %>%
  rename(Kingdom = taxonomy1, Phylum = taxonomy2, Class = taxonomy3, Order = taxonomy4, Family = taxonomy5, Genus = taxonomy6, Species = taxonomy7)

```

This chunk creates a dataframe for 5 diversity metrices from the OTU input file.  
```{r, include = TRUE}

# Calculate diversity indices and create table 
################################################### 
div <- matrix()

Shannon <- diversity(x = OTU, index = "shannon")
Richness <- rowSums(OTU > 0) 
Evenness <- Shannon/log(Richness)
Simpsons <- diversity(OTU, "simpson")
iSimpsons <- diversity(OTU, "inv")

div$Shannon <- Shannon
div$Richness <- Richness
div$Evenness <- Evenness
div$Simpsons <- Simpsons
div$iSimpsons <- iSimpsons
div <- as.data.frame(div) %>%
  rownames_to_column(var = "SampleID") %>%
  select("SampleID", "Shannon", "Richness", "Evenness", "Simpsons", "iSimpsons") 
```

After testing if the numerical variables are normally distributed, this chunk takes the log of the variable and adds one (to avoid neg. values). The output is a metadata file with the categorical variables of interest, variables that were normally distributed without transformation, as well the variables that needed transformation. 
```{r, include = TRUE}
# Log transform non-normal variables
meta_data_log <- metadata %>% 
  mutate(logElev = log(Elevation_m) + 1) %>%
  mutate(logMax_Dep = log(Max_Depth_m) + 1) %>%
  mutate(logTDP = log(TDP) + 1) %>%
  mutate(logDOC = log(DOC) + 1) %>%
  mutate(logChl_a = log(Chl_a) + 1) %>%
  mutate(logTempS = log(Surface_Temp_c) + 1) %>%
  select("SampleID", "DOMsource", "LakeType", "Sample_Location", "logElev", "logMax_Dep", "logTempS", "SSRatio", "logTDP", "NH3", "logDOC", "pH", "logChl_a")
head(meta_data_log)

```

This chunk merges the last metadata file with the dataframe of diversity indices and writes a csv to disk. 
```{r, include = TRUE}
# Merge diversity indices with metadata and save to file. 
div_data <- merge(div, meta_data_log, by = "SampleID")
write.csv(div_data, paste0(results.fp, "/DiversityIndices.csv"))
```

This chunk defines the dependent variable of interest and creates the user-friendly title for plotting. 
```{r, include = TRUE}
# Define data
data <- div_data

# Define the dependent variable
dvar_name <- "Shannon"

if (dvar_name == "Richness") {
dvar <- data$Richness
dvar_text <- "Richness (Number of Species)"
}
if (dvar_name == "Evenness") {
dvar <- data$Evenness
dvar_text <- "Pielou's Evenness Index"
}
if (dvar_name == "Shannon") {
dvar <- data$Shannon
dvar_text <- "Shannon Diversity Index (H)"
}
if (dvar_name == "Simpsons") {
dvar <- data$Simpsons
dvar_text <- "Simpson's Diversity Index"
}
if (dvar_name == "iSimpsons") {
dvar <- data$iSimpsons
dvar_text <- "Inverse Simpson's Diversity Index"
}

paste0("The dependent variable we are interested in is: ", dvar_text)
```

This chunk defines the independent variable of interest, creates the user-friendly variable title for plotting, user-friendly treatment names, and defines the colors associated with that variable.
```{r, include = TRUE}
# Define the independent variable
ivar_name <- "LakeType"

if (ivar_name == "DOMsource") {
  ivar <- data$DOMsource
  ivar_text <- "DOM Source"
  legend.labels <- c("Algae-derived", "Terrestrially-derived")
  colors <- c("#99A633", "#936226") # Algal: olive green, Terrestrial: brown
}
if (ivar_name == "LakeType") {
  ivar <- data$LakeType
  ivar_text <- "Lake Type"
  legend.labels <- c("Alpine Lakes", "Subalpine Lakes")
  colors <- c("#8acfcf", "#99A633") # Alpine: light teal, Subalpine: light olive green
}
if (ivar_name == "Sample_Location") {
  ivar <- data$Sample_Location
  ivar_text <- "Sample Location"
  legend.labels <- c("Lake Surface", "Lake Bottom")
  colors <- c("#146891", "#b4e4f0") # Bottom: dark blue, Surface: light blue
}
if (ivar_name == "SSRatio") {
  ivar <- data$SSRatio
  ivar_text <- "SS Ratio"
}
if (ivar_name == "logElev") {
  ivar <- data$logElev
  ivar_text <- "Elevation (m)"
}
if (ivar_name == "logMax_Dep") {
  ivar <- data$logMax_Dep
  ivar_text <- "Max Depth (m)"
}
if (ivar_name == "logTempS") { 
  ivar <- data$logTempS
  ivar_text <- "Surface Temp (C)"
}

paste0("The dependent environmental variable we are looking at is: ", ivar_text)
```

This chunk runs a simple linear regression. Use if both independent and dependent variable are continuous. 
```{r, include = TRUE}
sum <- summary(lm(data = data, formula = dvar ~ ivar))

P_regression <- round(sum$coefficients[,4], digits = 3)[[2]]

if (P_regression < 0.05) {
  paste0(dvar_text, " differs by ", ivar_text, " (p = ", P_regression, ")")
} else { 
  paste0(dvar_text, " does not differ by ", ivar_text, " (p = ", P_regression, ")")}

```

This chunk runs a student's t-test. Use if both independent variable is continuous (with two treatments) and the dependent variable is categorical. 
```{r, include = TRUE}
t <- t.test(dvar ~ ivar, data = data)
P_ttest <- round(t$p.value, digits = 3)

if (P_ttest < 0.05) {
  paste0(dvar_text, " differs by ", ivar_text, " (p = ", P_ttest, ")")
} else { 
  paste0(dvar_text, " does not differ by ", ivar_text, " (p = ", P_ttest, ")")}
```

Plotting boxplots for categorical independent variable. This chunk defines the format of the plot. 
```{r, include = TRUE}
# What font do you want to use throughout? 
font.family <- "Avenir"

# Add title information
title <- paste0(dvar_text, " by ", ivar_text) # or replace with new title. 
title.size <- 18 # Font size
title.face <- "bold" # Font face
title.color <- "black" # Font color
title.family <- font.family # Font will be the same as defined above unless changed here. 

# Define the x and y axes labels. 
ytitle <- dvar_text

xtitle <- ivar_text
x.title.hjust <- 0
x.title.vjust <- -1

x.text.angle <- 0
x.text.hjust <- 0.5
x.text.vjust <- 1

# Do you want to change the colors that were defined above for the variable? If so, change the colors that were defined earlier based on the variable. 
colors <- colors

# Define legend attributes
legend.title <- ivar_text
legend.title.size <- 13
legend.text.size <- 11 # Legend text size
legend.labels <- legend.labels

# Write the p-value and R2 on the plot
stats_vjust <- 1 # vertical location
stats_hjust <- -3 # horizontal location

# What do you want the outliers to look like? 
outlier.size <- 1
outlier.colour <- "grey30"
```

This chunk uses the specifications from above to plot the figure. 
```{r, include = TRUE}

plot <- ggplot(data, aes(x = ivar, y = dvar, fill = ivar)) +  
  xlab(xtitle) +
  ylab(ytitle) +
  ggtitle(title) +
  geom_boxplot(outlier.size = outlier.size, outlier.colour = outlier.colour) + 
  theme(text = element_text(family = font.family),
        axis.title.x = element_text(vjust = x.title.vjust),
        legend.title = element_text(size = legend.title.size),
        legend.text = element_text(size = legend.text.size),
        plot.title = element_text(family = title.family, 
                                  size = title.size, 
                                  face = title.face, 
                                  color = title.color))   +
  scale_fill_manual(values = colors, name = legend.title, labels = legend.labels)  

plot
```

This chunk saves the plot to the figures folder. 
```{r, include = TRUE}
# Save figure to disk
figure.type <- ".png" 
filename <- paste0(dvar_name, " by ", ivar_name, "_", primer)
width = 6
height = 5

ggsave(plot = plot, 
       filename = paste0(figures.fp, "/", filename, figure.type), 
       height = height, 
       width = width)
```

